### DISCUSSION TITLE
NoDecodeDelegateForThisImageFormat and MissingAnImageFilename error for composite tile command.

### DISCUSSION THREAD

**Discussion by rsyqvthv** [ImageMagick/ImageMagick#2692] [General]
I am new in ImageMagick, I know my question must stupid, but I really get stuck, and can't find any solution after 6 hours research.

What I would like to do is just tile the 'copyright' text on a image (`input.jpg`) in Windows Powershell

I borrowed piece of code from net, adjust bit to suitable my needs, the `input.jpg` in current folder:

```
magick convert -size 160x100 xc:none -fill grey -gravity NorthWest -draw "rotate 30 text 20,20 'Copyright'" miff:- | magick composite -tile - .\input.jpg output.jpg
```

but it always shows me the error:

```
composite: NoDecodeDelegateForThisImageFormat `' @ error/constitute.c/ReadImage/562.
composite: MissingAnImageFilename `output.jpg' @ error/composite.c/CompositeImageCommand/1629.
```

I don't know what's problem of this, as the similar code was works on F-xVAAqzNOM@youtube

**Comment by snibgo** [comment]
What version of IM?

Your command works fine for me at the command line, with IM v7.0.8-64 on Windows 8.1:
```
f:\web\im>%IMG7%magick -size 160x100 xc:none -fill grey -gravity NorthWest -draw "rotate 30 text 20,20 'Copyright'"  miff:- | %IMG7%magick composite -tile - t.png t2.png
```
Does it work for you at the command line?

If so, then I would suspect a problem with how you do this in Powershell. Perhaps a problem with quotes within quotes.

**Reply by rsyqvthv** [reply]
> What version of IM?

`Version: ImageMagick 7.0.10-30 Q8 x64 2020-09-19 http://www.imagemagick.org`

>Does it work for you at the command line?

I did ran them at the command line. windows 10, powershell 5.1.17763.316

> Perhaps a problem with quotes within quotes.

I did use the exactly same code I post on the OP, I have no idea if the quotes was wrong. :(

**Comment by fmw42** [comment]
Can you post your input and copyright image so we can use them.  For Imagemagick 7, use magick, not magick convert and not convert. 

Here is another way (using in memory mpr: images) that does not require the use of piping.

![lena](https://user-images.githubusercontent.com/13458218/95364331-8df13580-0885-11eb-96da-e2d52d5af02f.png)


 `magick -size 50x -background none -fill black label:"CopyRight" -rotate 30 -bordercolor none -border 20 -write mpr:tile +delete lena.png +write mpr:img -fill mpr:tile  -draw 'color 0,0 reset' mpr:img +swap -compose over -composite x.png`


![x](https://user-images.githubusercontent.com/13458218/95364353-95b0da00-0885-11eb-8183-3a5ed1014a17.png)

**Comment by rsyqvthv** [comment]
> Can you post your input and copyright image so we can use them. 

I didn't have copyright image, only the `input.jpp` here: 
![input](https://user-images.githubusercontent.com/2230369/95366699-b00da680-08cb-11eb-91ae-09477bc726cb.jpg)

> For Imagemagick 7, use magick, not magick convert and not convert.

Oh, yes, tried magick only command: 

```
 magick -size 160x100 xc:none -fill grey -gravity NorthWest -draw "rotate 30 text 20,20 'Copyright'" miff:- | magick -tile - .\input.jpg output.jpg
```

seems it kills one error, but leave another one remaining:

```
magick.exe: NoDecodeDelegateForThisImageFormat `' @ error/constitute.c/ReadImage/562.
```


> Here is another way (using in memory mpr: images) that does not require the use of piping.

your command works perfectly~~! thank you, but I have to spend bit more time to understand it. 
looks like mine one is bit simpler for newbie as me, if it works. I am curious what's stop it from working?

**Reply by fmw42** [reply]
Perhaps you OS or Powershell does not allow you to pipe.

**Reply by rsyqvthv** [reply]
any simple way to testing if pipe works at my powershell?

```
ls *.mp4 | %{$_.BaseName}
```

above is works.

Your guess is make me think of the pipe `|` in the OP Powershell command is same as the pipe in Magick, 
so how Powershell can tell which one is for Powershell command and which one is only for Magick?

Maybe this is the issue?

**Comment by fmw42** [comment]
Here is how to do that your way. You have to create the tile image of the size of the input, then composite it over the input.  See https://imagemagick.org/Usage/canvas/#tile

 `magick -size 160x100 xc:none -fill white -gravity NorthWest -draw "rotate 30 text 20,20 'Copyright'" miff:- | im7 magick input.jpg \( +clone -tile - -draw "color 0,0 reset" \) -compose over -composite output.jpg`

![output](https://user-images.githubusercontent.com/13458218/95375878-1deaab80-0895-11eb-8e6f-e3f1cd5b9471.jpg)

**Reply by rsyqvthv** [reply]
thank you @fmw42 your code here produce lost of errors in my Powershell. 

I saw `\(....\)` in the segment, I thought you are in *nix system, I tried my code of OP in Bash (windows version), which is working perfectly.

---

I do suspect, as you said above was the point:

> Perhaps you OS or Powershell does not allow you to pipe [when I use magick, it must be mess the Powershell pipe and Magick pipe somehow]

carefully red the [document ](https://imagemagick.org/script/command-line-processing.php) of Image Magick, especially the navy wording on the page, it mentioned __escape word__. 

so I tried my code of OP with piple escaped, (put a powershell escape char backtick ` before pipe | )

```
magick -size 160x100 xc:none -fill grey -gravity NorthWest -draw "rotate 30 text 20,20 'Copyright'" miff:- `| magick -tile - .\input.jpg output.jpg
```

but it stuck when I press enter, command keep running without any output till I press ctrl+c to cancel. 

---

@fmw42 

as I red the document, I find Magick support another way to do pipe, without using `|`, it mentions in [input filename](https://imagemagick.org/script/command-line-processing.php#input) blow the __STDIN, STDOUT, and file descriptors__ section.

I tried using this concept to adjustment my OP code bit as: 

```
magick -size 160x100 xc:none -fill grey -gravity NorthWest -draw "rotate 30 text 20,20 'Copyright'" miff:-  -tile fd:3 .\input.jpg output.jpg
```

it works same as I escaped pipe above, nothing output, have to ctrl+c to cancel.  

Any suggestion of how to use the `fd` concept to make my OP code works?


Thank you for your helping.

**Reply by fmw42** [reply]
Sorry, I have never used fd. Perhaps someone else cn inform you. But if you are using Powershell, the syntax may be totally different. You need to refer to Powershell docs.

**Reply by rsyqvthv** [reply]
thanks @fmw42 

currently I am using the code you gave me yesterday, 
I've research whole day today to figure out what's the meaning of this code, 
and learned lot from it, which isn't that complex as I feel when first met it.

thank you again.

---

hope someday I can figure out how Powershell handling the Magick pipe properly. 

:)

**Reply by fmw42** [reply]
See https://www.goingforawalkwiththedogs.com/2017/09/calling-image-magick-from-powershell.html and https://gist.github.com/curtisgibby/ebac612657d6ef71a32fe03e75bd07cf and https://issue.life/questions/47292991
